### ISSUE TITLE
Improve error messages for duplicate identifiers

### ISSUE THREAD

**Issue by mhegazy** [microsoft/TypeScript#529]
We need to split the "Duplicate identifier" error in to a suite of more detailed errors about various types of illegal merges. "Duplicate identifier" is too general. Some examples of cases that warrant their own message:
- Would be a legal merge (like module & module) but exports don't match
- Fundule in the wrong order
- Clodule in the wrong order
- Fundule in different files
- Clodule in different files
- Interface and enum can't merge
- Interface and class can't merge
- Class and enum can't merge
- Function and enum can't merge

Basically every illegal cell in the matrix of merges should have a separate error message (or some of them can be grouped together in terms of value/type/namespace space). For example:

Class cannot merge with any other type

**Comment by robertknight** [comment]
Additional related request, it would be useful if the message included the source location of the existing identifier.

**Comment by mhegazy** [comment]
@robertknight we used to do that in the 1.0 time-frame, but then the error message was too long and not easily piped through IDEs and editors to allow jumping to the location; instead, the new behavior is to error on all instances of the duplicate definitions, this way you can chose which one to jump to.

**Comment by brandonbloom** [comment]
> the new behavior is to error on all instances of the duplicate definitions, this way you can chose which one to jump to.

Unfortunately, if the duplicate identifier is not in your code, for example if you do `class Node { ...`, then you can't find the other definition easily. Presumably "Node" is colliding with something from the DOM typings.

**Comment by kitsonk** [comment]
The compiler outputs all locations...

```
$ cat test.ts
class Node { }
$ tsc test.ts
/node_modules/typescript/lib/lib.d.ts(12324,11): error TS2300: Duplicate identifier 'Node'.
/node_modules/typescript/lib/lib.d.ts(12376,13): error TS2300: Duplicate identifier 'Node'.
test.ts(1,7): error TS2300: Duplicate identifier 'Node'.
```

If you are in vscode, the quick info does not provide all locations, but if you were to comment out the offending code and just write:

```ts
Node;
```

The quick info would provide you with the definition locations you could navigate to.

**Comment by NaridaL** [comment]
- [ ]     Would be a legal merge (like module & module) but exports don't match
```
namespace FF {
    export function foo() { // error here, not on FF "Duplicate function implementation"
        return 0
    }
}
namespace FF {
    export function foo() {
        return "2"
    }
}
```
which seems appropriate...?
- [X]     Fundule in the wrong order
- [X]     Clodule in the wrong order
>A namespace declaration cannot be located prior to a class or function with which it is merged.
- [X]     Fundule in different files 
- [X]     Clodule in different files
>A namespace declaration cannot be in a different file from a class or function with which it is merged.
- [ ]     Interface and class can't merge
No error at all currently: https://www.typescriptlang.org/play/#src=interface FF {%0D%0A foo()%3A void%0D%0A}%0D%0Aclass FF {%0D%0A x%3A string%0D%0A}

The following are still an issue:
- [ ]     Interface and enum can't merge
- [ ]     Class and enum can't merge
- [ ]     Function and enum can't merge

**Comment by RyanCavanaugh** [comment]
## Unable to Reproduce ❌

I attempted to reproduce this issue but could not confirm it still exists.

Test case used:
**namespace1.ts:**
```typescript
namespace A {
    export function foo() {
        return 0;
    }
}
```

**namespace2.ts:**
```typescript
namespace A {
    export function foo() {
        return "2";
    }
}
```

**interface_enum.ts:**
```typescript
interface IExample {
    foo(): void;
}

enum IExample {
    A,
    B,
}
```

**class_enum.ts:**
```typescript
class ExampleClass {
    x: number = 0;
}

enum ExampleClass {
    A,
    B,
}
```

**function_enum.ts:**
```typescript
function exampleFunc() {
    console.log("Hello");
}

enum exampleFunc {
    A,
    B,
}
```

**tsconfig.json:**
```json
{
    "compilerOptions": {
        "strict": true
    }
}
```


**Command:** `tsc -p .`
**TypeScript Version:** Version 5.9.0-dev.20250722
**Exit Code:** 2

<details>
<summary>Compiler Output</summary>

**Stdout:**
```
C:/Users/ryanca/AppData/Local/Temp/repro-check-KYjRhX/class_enum.ts(1,7): error TS2567: Enum declarations can only merge with namespace or other enum declarations.
C:/Users/ryanca/AppData/Local/Temp/repro-check-KYjRhX/class_enum.ts(5,6): error TS2567: Enum declarations can only merge with namespace or other enum declarations.
C:/Users/ryanca/AppData/Local/Temp/repro-check-KYjRhX/function_enum.ts(1,10): error TS2567: Enum declarations can only merge with namespace or other enum declarations.
C:/Users/ryanca/AppData/Local/Temp/repro-check-KYjRhX/function_enum.ts(5,6): error TS2567: Enum declarations can only merge with namespace or other enum declarations.
C:/Users/ryanca/AppData/Local/Temp/repro-check-KYjRhX/interface_enum.ts(1,11): error TS2567: Enum declarations can only merge with namespace or other enum declarations.
C:/Users/ryanca/AppData/Local/Temp/repro-check-KYjRhX/interface_enum.ts(5,6): error TS2567: Enum declarations can only merge with namespace or other enum declarations.
C:/Users/ryanca/AppData/Local/Temp/repro-check-KYjRhX/namespace1.ts(2,21): error TS2393: Duplicate function implementation.
C:/Users/ryanca/AppData/Local/Temp/repro-check-KYjRhX/namespace2.ts(2,21): error TS2393: Duplicate function implementation.

```

**Stderr:**
```
(empty)
```
</details>

**What you observed in the bug report:** The error messages for duplicate identifiers and illegal merges were too general, making it difficult to understand the specific issue or easily find the source of the conflict.
**What you expected:** Error messages should be specific to the type of illegal merge occurring, and should include source locations for existing identifiers to aid in debugging.
**What I observed when running tsc:** Error messages are specific to the type of illegal merge and include source locations, such as "Enum declarations can only merge with namespace or other enum declarations" and "Duplicate function implementation," with exact file and line references.

My reasoning:
> The bug reported was that the error messages for duplicate identifiers and illegal merges were too generic, making it difficult to understand the specific issue. Users expected more detailed error messages explaining the nature of the conflict.

The user-reported behavior described a circumstance where error messages were not specific enough, causing difficulty in identifying the type of illegal merge happening, and a lack of location information for the existing identifier.

In the current compiler output, we see more specific error messages indicating issues like "Enum declarations can only merge with namespace or other enum declarations" and "Duplicate function implementation," which directly address specific cases of illegal merges. Moreover, the error messages now include source locations for both the conflicting instances, making it easier to identify the issues.

Thus, the current observed behavior reflects an improvement in the specificity of error messages, aligning with what the users wanted. The bug has been fixed as the problematic behavior reported by users is no longer present, and the expected behavior of detailed error messaging is now observed.


---

**Analysis Result:**
```json
{
  "bug_is_present": "no",
  "files": [
    {
      "filename": "namespace1.ts",
      "content": "namespace A {\n    export function foo() {\n        return 0;\n    }\n}"
    },
    {
      "filename": "namespace2.ts",
      "content": "namespace A {\n    export function foo() {\n        return \"2\";\n    }\n}"
    },
    {
      "filename": "interface_enum.ts",
      "content": "interface IExample {\n    foo(): void;\n}\n\nenum IExample {\n    A,\n    B,\n}"
    },
    {
      "filename": "class_enum.ts",
      "content": "class ExampleClass {\n    x: number = 0;\n}\n\nenum ExampleClass {\n    A,\n    B,\n}"
    },
    {
      "filename": "function_enum.ts",
      "content": "function exampleFunc() {\n    console.log(\"Hello\");\n}\n\nenum exampleFunc {\n    A,\n    B,\n}"
    },
    {
      "filename": "tsconfig.json",
      "content": "{\n    \"compilerOptions\": {\n        \"strict\": true\n    }\n}"
    }
  ],
  "user_reported_behavior": "The error messages for duplicate identifiers and illegal merges were too general, making it difficult to understand the specific issue or easily find the source of the conflict.",
  "expected_behavior": "Error messages should be specific to the type of illegal merge occurring, and should include source locations for existing identifiers to aid in debugging.",
  "current_observed_behavior": "Error messages are specific to the type of illegal merge and include source locations, such as \"Enum declarations can only merge with namespace or other enum declarations\" and \"Duplicate function implementation,\" with exact file and line references.",
  "command_line": "tsc -p .",
  "tsc_version": "Version 5.9.0-dev.20250722",
  "compiler_output": {
    "stdout": "C:/Users/ryanca/AppData/Local/Temp/repro-check-KYjRhX/class_enum.ts(1,7): error TS2567: Enum declarations can only merge with namespace or other enum declarations.\r\nC:/Users/ryanca/AppData/Local/Temp/repro-check-KYjRhX/class_enum.ts(5,6): error TS2567: Enum declarations can only merge with namespace or other enum declarations.\r\nC:/Users/ryanca/AppData/Local/Temp/repro-check-KYjRhX/function_enum.ts(1,10): error TS2567: Enum declarations can only merge with namespace or other enum declarations.\r\nC:/Users/ryanca/AppData/Local/Temp/repro-check-KYjRhX/function_enum.ts(5,6): error TS2567: Enum declarations can only merge with namespace or other enum declarations.\r\nC:/Users/ryanca/AppData/Local/Temp/repro-check-KYjRhX/interface_enum.ts(1,11): error TS2567: Enum declarations can only merge with namespace or other enum declarations.\r\nC:/Users/ryanca/AppData/Local/Temp/repro-check-KYjRhX/interface_enum.ts(5,6): error TS2567: Enum declarations can only merge with namespace or other enum declarations.\r\nC:/Users/ryanca/AppData/Local/Temp/repro-check-KYjRhX/namespace1.ts(2,21): error TS2393: Duplicate function implementation.\r\nC:/Users/ryanca/AppData/Local/Temp/repro-check-KYjRhX/namespace2.ts(2,21): error TS2393: Duplicate function implementation.\r\n",
    "stderr": "",
    "exit_code": 2
  },
  "reasoning": "The bug reported was that the error messages for duplicate identifiers and illegal merges were too generic, making it difficult to understand the specific issue. Users expected more detailed error messages explaining the nature of the conflict.\n\nThe user-reported behavior described a circumstance where error messages were not specific enough, causing difficulty in identifying the type of illegal merge happening, and a lack of location information for the existing identifier.\n\nIn the current compiler output, we see more specific error messages indicating issues like \"Enum declarations can only merge with namespace or other enum declarations\" and \"Duplicate function implementation,\" which directly address specific cases of illegal merges. Moreover, the error messages now include source locations for both the conflicting instances, making it easier to identify the issues.\n\nThus, the current observed behavior reflects an improvement in the specificity of error messages, aligning with what the users wanted. The bug has been fixed as the problematic behavior reported by users is no longer present, and the expected behavior of detailed error messaging is now observed."
}
```
